Ignore version numbers in branch names that are not release branches.

bert2 opened the pull request:
This fixes #1536.

~~Introduces a new config switch in the [`ignore` section](https://github.com/bert2/GitVersion/blob/bugfix/issue-1536-ignore-versions-in-branch-names/docs/configuration.md#ignore) called [`non-release-branches`](https://github.com/bert2/GitVersion/blob/bugfix/issue-1536-ignore-versions-in-branch-names/docs/configuration.md#non-release-branches). Setting it to `true` will ignore versions coming from branch names that are not [release branches](https://github.com/bert2/GitVersion/blob/bugfix/issue-1536-ignore-versions-in-branch-names/docs/configuration.md#is-release-branch).~~

Branches that are not [release branches](https://github.com/bert2/GitVersion/blob/bugfix/issue-1536-ignore-versions-in-branch-names/docs/configuration.md#is-release-branch) will not be considered for version calculation anymore.

Works with versions coming from branch names in merge commits and from the currently checked out branch.

Also works for remote-tracking branches as long as their remote is `origin`.

**Additional notes:**

* ~~There is some duplication in the strategies `MergeMessageBaseVersionStrategy` and `VersionInBranchNameBaseVersionStrategy` that I didn't know where to extract it to.~~
    * The `Config` class seemed like a good place.
* ~~The check if a branch is a release branch fails when a remote branch is merged due to the `origin/` prefix in the merge commit message: `"Merge remote-tracking branch 'origin/release/0.8.0' into develop/master"`. Should such a merge be considered a merge of a release branch? Please let me know and I'll ammend the PR.~~
    * Fixed by removing `"refs/remotes/<remote-name>/"` from the branch name if present. For instance `"refs/remotes/origin/release/2.0.0"` will be stripped to it's friendly name without the remote, i.e. `"release/2.0.0"`.
* ~~I'm not sure how some branches should be handled. Originally the branch `Particular/release-1.0.0` would always yield the version `1.0.0`. With this PR the above branch name will only be considered for version calculation when `Particular` also is a remote reference of the repository. Is that assumption correct?~~
    * Only merges of remote-tracking release branches coming from `origin` can yield versions. All other remotes are ignored.

bert2 commented:
Something ain't right with Travis... He failed the first build with a random error in the linux job while the osx job went through.
And now he just failed the osx job with another random error, but the linux job passed?

asbjornu commented:
@bert2: I ran the build again, so it's not successful. If you can `rebase` against the latest `master`, this LGTM. Thanks! 😃

asbjornu commented:
As #1422 is close to being merged and we discuss incrementing major with it, I think we should reverse the current behavior and ignore the version number in branches that aren't release branches by default. I'm not even sure we need a config for enabling the current (imho unintuitive) behavior. Thoughts, @bert2?

asbjornu commented:
@bert2: Yes, whether the branch is remote or local shouldn't matter in its consideration for release classification, I think. So remote or local are just the same; a release branch is a release branch no matter where it lives.

Just take this up when you have time for it and when you do, I suggest you rebase on `master` (I hope #1422 will be merged by then), remove the config and reverse the logic so it applies by default.

bert2 commented on src/GitVersionCore.Tests/Mocks/MockRepository.cs:
```diff
@@ -1,5 +1,8 @@
 using System;
 using System.Collections.Generic;
+
+using GitVersionCore.Tests.Mocks;
```
The `Mock*` classes are all without namespace. I accidentally created the new ones with the sub-namespace `Mock`, but will fix this later.

bert2 commented:
Hey @asbjornu, I changed the PR as suggested and updated the description above. Three tests are still failing, because I'm not sure what the expected behavior should be. Please see [this comment](https://github.com/GitTools/GitVersion/pull/1541/files/9fd376c8110d7ab3e87b310bf781be9f4f0e3532#r243334099).

asbjornu commented on src/GitVersionCore.Tests/VersionCalculation/Strategies/MergeMessageBaseVersionStrategyTests.cs:
```diff
@@ -30,38 +32,25 @@ public void ShouldNotAllowIncrementOfVersion()
         [TestCase("Merge branch 'Release-v0.2.0'", true, "0.2.0")]
         [TestCase("Merge branch 'Release-v2.2'", true, "2.2.0")]
         [TestCase("Merge remote-tracking branch 'origin/release/0.8.0' into develop/master", true, "0.8.0")]
-        public void AssertMergeMessage(string message, bool isMergeCommit, string expectedVersion)
```
Yes, I agree; `Particular` should be considered a remote ref that should not affect the version number by default. This is a breaking change we're introducing here and I think it's fine.

bert2 replied:
@asbjornu, I'm not sure I understand. Currently _all_ remote refs of a repo are considered when shortening a branch name to its friendly name.

So when `Particular` is listed as `[remote "Particular"]` in the git config of the repo then the branch name `Particular/release-1.0.0` will be reduced to `release-1.0.0` and thus be considered a release branch affecting the version.

This is done by the new method [`AsBranchNameWithoutRemote()`](https://github.com/GitTools/GitVersion/pull/1541/files/9fd376c8110d7ab3e87b310bf781be9f4f0e3532#diff-82df1f328d9f7f5a1f9b55d7d88d7e38).

If I understand you correctly then only "special" remotes like `origin` and maybe `upstream` should be allowed to affect the version. In that case the above methods needs to be changed.

bert2 replied:
Ok, nice. This makes things a little easier then.

asbjornu commented:
Excellent. We now just need to figure out #1550 and we should be good to go. Thanks for your patience and all the work you’ve put into this! :pray:

richardgavel commented:
Any chance this will be able to be merged soon?

asbjornu commented:
Now that #1581 is merged, I think we're in a better place wrt. #1550. What do you say, @arturcic? Do we have a working product again? I'm sorry I don't have time to test this myself at the moment. 😞

arturcic commented:
More or less yes, we still use older version of Libgit2sharp, and I have integrated the code from GitTools.Core to use directly the Libgit2Sharp library. I'm planning to upgrade it and fix the packaging. For now the current version relies on the workaround to explicitly install the needed libgit2-library on linux

arturcic commented:
I guess we can integrate some PRs that improve the code, as the changes I'm planning are more related to build/packaging

asbjornu commented:
@arturcic, Ok, sounds good. The only thing missing from #1422 then is https://github.com/GitTools/GitVersion/pull/1422#issuecomment-449349025. I guess any commit is as good as any for a `5.0.0-alpha.1` tag?

arturcic commented:
> @arturcic, Ok, sounds good. The only thing missing from #1422 then is [#1422 (comment)](https://github.com/GitTools/GitVersion/pull/1422#issuecomment-449349025). I guess any commit is as good as any for a `5.0.0-alpha.1` tag?

Agree

asbjornu commented:
Since #1422 marked the beginning of this whole debacle, I tagged cf2ccfa2db2f4628259fc170cfba8de0f040c551 with `5.0.0-alpha.1`.
